src/lib/masterSpan/index.js

```diff
@@ -0,0 +1,159 @@
+const Config = require('../config')
+const partition = 'endpoint-cache'
+const clientOptions = { partition }
+let client
+
+const sleep = (ms) => {
+  return new Promise(resolve => {
+    setTimeout(resolve, ms)
+  })
+}
+
+const initializeCache = async () => {
+  client = new Catbox.Client(CatboxMemory, clientOptions)
+  await client.start()
+  return true
+}
+
+const cacheSpanContext = async (spanContext, state, content) => {
+  const { traceId, parentSpanId, tags, service, spanId } = spanContext
+  const { status, code } = state
+  const newSpanId = crypto.randomBytes(8).toString('hex')
+
+  const masterKey = {
+    segment: `${Config.CACHE.segment}-${traceId}`,
+    id: newSpanId
+  }
+
+  const parentKey = {
+    segment: `${Config.CACHE.segment}-${traceId}`,
+    id: parentSpanId
+  }
+
+  const key = {
+    segment: `${Config.CACHE.segment}-${traceId}`,
+    id: spanId
+  }
+
+  try {
+    if (!parentSpanId) {
+      const newChildContext = merge({ parentSpanId: newSpanId }, { ...spanContext })
+      const newMasterSpanContext = merge({ tags: { ...tags, masterSpan: newSpanId } }, { ...spanContext }, { spanId: newSpanId, service: `master-${tags.transactionType}` })
+      await client.set(masterKey, { spanContext: newMasterSpanContext, state, content }, Config.CACHE.ttl)
+      return cacheSpanContext(newChildContext, state)
+    }
+    const parentSpan = !!parentSpanId && (await client.get(parentKey)).item.spanContext
+    const parentSpanTags = parentSpan ? parentSpan.tags : null
+    const errorTags = ('errorCode' in parentSpanTags) ? { errorCode: parentSpanTags.errorCode } : null
+    const masterTags = ('masterSpan' in parentSpanTags) ? { masterSpan: parentSpanTags.masterSpan } : null
+    const tagsToApply = merge({ ...errorTags }, { tags: { ...tags, ...masterTags } })
+
+    const isLastSpan = () => {
+      const { transactionAction, transactionType } = tags
+      const isError = (!!(code) && status === EventStatusType.failed) || errorTags
+      for (const criteria of Config.SPAN.END_CRITERIA[transactionType]) {
+        if (criteria[transactionAction]) {
+          if (('isError' in criteria[transactionAction]) && criteria[transactionAction].isError && isError) {
+            return true
+          } else if (!('isError' in criteria[transactionAction]) && criteria[transactionAction].service === service) {
+            return true
+          }
+        }
+      }
+      return false
+    }
+
+    await client.set(key, { spanContext: merge(tagsToApply, { ...spanContext }), state, content }, Config.CACHE.ttl)
+
+    if (isLastSpan() || Config.SPAN.END_CRITERIA.exceptionList.includes(service)) { // TODO remove exceptionList when all traces have right tags
+      const trace = await createTraceArray(traceId, spanId)
+      const result = await createTrace(trace)
+      return result
+    }
+    return true
+  } catch (e) {
+    Logger.error(e)
+    throw e
+  }
+}
+
+const createTraceArray = async (traceId, lastSpanId) => {
```

Comment:
Let's discuss an alternative approach to this, rather use the incoming traces to kick-off the 'recording' process.

Reply:
:+1: